filesystem/etc/calico/confd/templates/bird.cfg.template

```diff
@@ -59,43 +59,120 @@ template bgp bgp_template {
+  {{- if selectsme $data.node_selector}}
+
+# BGPPeer {{$data.name}}
+# {{$data}}
 
+    {{- if ne "" $data.ip}}
 
-# ------------- Global peers -------------
-{{if ls "/global/peer_v4"}}
-{{range gets "/global/peer_v4/*"}}{{$data := json .Value}}
-{{$nums := split $data.ip "."}}{{$id := join $nums "_"}}
-# For peer {{.Key}}
-protocol bgp Global_{{$id}} from bgp_template {
+# Peer with IP {{$data.ip}}
+      {{- $nums := split $data.ip "."}}{{$id := join $nums "_"}}
+      {{- $node_cid_key := printf "/host/%s/cluster_id" (getenv "NODENAME")}}{{if exists $node_cid_key}}{{$cluster_id := getv $node_cid_key}}
+        {{- if ne "" ($cluster_id)}}
+protocol bgp Node_{{$id}} from bgp_template {
+  # We're a RR but our peer isn't - act as a client. TODO: Support peering with other RRs explicitly.
   neighbor {{$data.ip}} as {{$data.as_num}};
+  rr client;
+  rr cluster id {{$cluster_id}};
 }
-{{end}}
-{{else}}# No global peers configured.{{end}}
+        {{- else}}
+protocol bgp Node_{{$id}} from bgp_template {
+  # We're not a RR. Our peer might be, but it doesn't change our config. 
+  neighbor {{$data.ip}} as {{$data.as_num}};
+}
+        {{- end}}
+      {{- end}}
+    {{- else if ne "" $data.peer_selector}}
+      {{- range matchingnodes $data.peer_selector}}
+        {{- $node_cid_key := printf "/host/%s/cluster_id" (getenv "NODENAME")}}
+        {{- $onode_cid_key := printf "/host/%s/cluster_id" .}}
+        {{- $onode_as_key := printf "/host/%s/as_num" .}}
+        {{- $onode_ip_key := printf "/host/%s/ip_addr_v4" .}}
+        {{- if exists $onode_ip_key}}
+          {{- $onode_ip := getv $onode_ip_key}}
+          {{- $nums := split $onode_ip "."}}{{$id := join $nums "_"}}
+          {{- if eq $onode_ip ($node_ip) }}
 
+# Calico node {{.}}
+# Skipping ourselves ({{$node_ip}})
+          {{- else}}
+            {{- if exists $node_cid_key}}{{$cluster_id := getv $node_cid_key}}
+              {{- if ne "" ($cluster_id)}}
+                {{- if exists $onode_cid_key}}{{$onode_cid := getv $onode_cid_key}}
+                  {{- if eq "" ($onode_cid)}}
 
-# ------------- Node-specific peers -------------
-{{$node_peers_key := printf "/host/%s/peer_v4" (getenv "NODENAME")}}
-{{if ls $node_peers_key}}
-{{range gets (printf "%s/*" $node_peers_key)}}{{$data := json .Value}}
-{{$nums := split $data.ip "."}}{{$id := join $nums "_"}}
-# For peer {{.Key}}
+# Calico node {{.}}
 protocol bgp Node_{{$id}} from bgp_template {
-  neighbor {{$data.ip}} as {{$data.as_num}};
+  # We're a RR but our peer isn't - act as a client.
+  neighbor {{$onode_ip}} as {{if exists $onode_as_key}}{{getv $onode_as_key}}{{else}}{{getv "/global/as_num"}}{{end}};
+  rr client;
+  rr cluster id {{$cluster_id}};
 }
-{{end}}
-{{else}}# No node-specific peers configured.{{end}}
-{{end}}{{/* End of IPv4 enable check */}}
+                  {{- else}}
+
+# Calico node {{.}}
+protocol bgp RR_{{$id}} from bgp_template {
+  # We're both RRs, don't act as a client.
```

Comment:
I think we may need to check if the RR is within our cluster or not - if it's not in our cluster, then we probably need to peer with it as a client. 

Need to double check this though.